src/Simulation/Simulators/QuantumSimulator/ClassicalControl.cs

```diff
@@ -0,0 +1,290 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using Microsoft.Quantum.Simulation.Core;
+
+namespace Microsoft.Quantum.Simulation.Simulators
+{
+    public partial class QuantumSimulator
+    {
+        #region Static Methods
+
+        private static void RunClause(ICallable op, OperationFunctor type, IQArray<Qubit>? ctrls)
+        {
+            switch (type)
+            {
+                case OperationFunctor.Body: op.Apply(QVoid.Instance); break;
+                case OperationFunctor.Adjoint: ((IAdjointable)op).Adjoint.Apply(QVoid.Instance); break;
+                case OperationFunctor.Controlled: ((IControllable)op).Controlled.Apply(ctrls); break;
+                case OperationFunctor.ControlledAdjoint: ((IUnitary)op).Controlled.Adjoint.Apply(ctrls); break;
+            }
+        }
+
+        private static QVoid ExecuteConditionalStatement(Result measurementResult, ICallable onZero, ICallable onOne, OperationFunctor type, IQArray<Qubit>? ctrls)
+        {
+            if (measurementResult == Result.Zero)
+            {
+                RunClause(onZero, type, ctrls);
+            }
+            else
+            {
+                RunClause(onOne, type, ctrls);
+            }
+            return QVoid.Instance;
+        }
+
+        private static QVoid ExecuteConditionalStatement(IQArray<Result> measurementResults, IQArray<Result> comparisonResults, ICallable onEqualOp, ICallable onNonEqualOp, OperationFunctor type, IQArray<Qubit>? ctrls)
+        {
+            if (AreEqual(measurementResults, comparisonResults))
+            {
+                RunClause(onEqualOp, type, ctrls);
+            }
+            else
+            {
+                RunClause(onNonEqualOp, type, ctrls);
+            }
+            return QVoid.Instance;
+        }
+
+        private static bool AreEqual(IQArray<Result> measurementResults, IQArray<Result> comparisonResults)
+        {
+            if (measurementResults == null || comparisonResults == null || measurementResults.Count != comparisonResults.Count)
+            {
+                // Defaulting to true is based on the QuantumProcessorBase.cs behavior, under StartConditionalStatement.
+                return true;
+            }
+
+            bool equal = true;
+            for (int i = 0; i < measurementResults.Count; i++)
+            {
+                if (measurementResults[i] != comparisonResults[i])
+                {
+                    equal = false;
+                    break;
+                }
+            }
+
+            return equal;
```

Comment:
Is there any reason this can't just return the result directly instead of setting the variable and breaking out of the loop?
```suggestion
            for (int i = 0; i < measurementResults.Count; i++)
            {
                if (measurementResults[i] != comparisonResults[i])
                {
                    return false;
                }
            }

            return true;
```

Reply:
done